src/reactA11yAccessibleHeadingsRule.ts

```diff
@@ -0,0 +1,194 @@
+            }
+        }
+        //super.visitVariableDeclaration(node);
+    }
+
+    protected visitJsxElement(node: ts.JsxElement): void {
+        const options: IValidationOptions = {
+            validateText: true,
+            validateOrder: true,
+            validateH1Count: true
+        };
+        this.validateElement(node, options);
+        super.visitJsxElement(node);
+    }
+
+    private validateElement(node: ts.JsxElement, options: IValidationOptions): void {
+        if (!node) {
+            return;
+        }
+        const openingElement: ts.JsxOpeningElement = node.openingElement;
+
+        if (VALID_HEADING_TYPES.has(openingElement.tagName.getText())) {
+            // Validate heading text
+            if (options.validateText) {
+                this.validateHeadingText(node);
+            }
+            // Validate heading elements are structured in increasing order of +1
+            if (options.validateOrder) {
+                this.validateHeadingsOrder(openingElement);
+            }
+            // Validate no more than 2 H1 headings
+            if (options.validateH1Count) {
+                this.validateH1HeadingCount(node);
+            }
+        }
+    }
+
+    private validateSelfClosingElement(node: ts.JsxSelfClosingElement): void {
+        if (!node) {
+            return;
+        }
+        if (VALID_HEADING_TYPES.has(node.tagName.getText())) {
+            this.addFailureAt(node.getStart(), node.getWidth(), EMPTY_HEADING_FAILURE_STRING);
+        }
+    }
+
+    private validateH1HeadingCount(node: ts.JsxElement): void {
+        if (node.openingElement.tagName.getText().toLowerCase() !== 'h1') {
+            return;
+        }
+        this.numberOfH1Headings += 1;
+        if (this.numberOfH1Headings > MAX_NUMBER_OF_H1_HEADINGS) {
+            this.addFailureAt(node.getStart(), node.getWidth(), BAD_NUMBER_H1_HEADING_FAILURE_STRING);
+        }
+    }
+
+    private validateHeadingsOrder(headingNode: ts.JsxOpeningElement): void {
+        const headingNumber: number = parseInt(headingNode.tagName.getText()[1], 10);
+        if (!this.lastKnownHeadingNumber) {
+            this.lastKnownHeadingNumber = headingNumber;
+            return;
+        }
+        if (headingNumber !== this.lastKnownHeadingNumber && headingNumber - 1 !== this.lastKnownHeadingNumber) {
+            this.addFailureAt(headingNode.getStart(), headingNode.getWidth(), BAD_ORDER_HEADING_FAILURE_STRING);
+        } else {
+            this.lastKnownHeadingNumber = headingNumber;
+        }
+    }
+
+    private validateHeadingText(headingNode: ts.JsxElement): void {
+        let headingText: string | undefined;
+        const innerNode = headingNode.children[0];
+        if (headingNode.children.length === 0) {
+            this.addFailureAt(headingNode.getStart(), headingNode.getWidth(), EMPTY_HEADING_FAILURE_STRING);
+        } else {
+            if (innerNode.kind === ts.SyntaxKind.JsxExpression) {
```

Comment:
Fun fact: for each `ts.SyntaxKind`, there should be a helper available from the `tsutils` package of the form `isSomeKind(node: ts.Node): node is SomeKind`. If you use those, you won't have to manually cast the node later on (e.g. `expressionNode` below); TypeScript will "narrow" the type of the node. Pretty nifty 😉 and lets you get away from extra variables and `.kin` checking logic such as what you're doing in `expressionNode`.

If there's any syntax kind that doesn't have a corresponding check in `tsutils`, add it in the `utils` file in this repo and file an issue on `tsutils` to add it.

Reply:
Awesome.